homeassistant/components/godice/config_flow.py

```diff
@@ -0,0 +1,52 @@
+"""Config flow for godice."""
+
+from __future__ import annotations
+
+from homeassistant.components.bluetooth import BluetoothServiceInfo
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_ADDRESS, CONF_NAME
+
+from .const import DOMAIN
+
+
+class GoDiceConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Setup GoDice discovered via Bluetooth."""
+
+    def __init__(self) -> None:
+        """Declare fields used later on device discovery."""
+        self._discovery_info: BluetoothServiceInfo | None = None
+
+    async def async_step_user(self, user_input=None) -> ConfigFlowResult:
+        """Handle the user step to pick discovered device."""
+        return self.async_abort(reason="auto_discovery_only")
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfo
+    ) -> ConfigFlowResult:
+        """GoDice discovery handler."""
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        self._discovery_info = discovery_info
+        return await self.async_step_discovery_confirm()
+
+    async def async_step_discovery_confirm(self, user_input=None) -> ConfigFlowResult:
+        """Show setup confirmation dialog and setup when confirmed."""
+        assert self._discovery_info is not None
```

Comment:
Same for the other assertions in the config_flow
```python
from typing import TYPE_CHECKING
```
```suggestion
        if TYPE_CHECKING:
            assert self._discovery_info is not None
```

Reply:
I don't think it is a good idea. Doing this way we don't learn from the checker but just make type checking pass while the checker watches but ignore its recommendations when it does not. Instead with assertion like that we declare explicitly that `None` is not expected and can fail early

Reply:
In practice that should never occur because self._discovery_info is always set a step before it is used. I had the same problem before and I was told you could use an assert only with `if TYPE_CHECKING`

Reply:
so if we're sure `self._discovery_info` should always be set, why to put `assert not None` under if, don't we assert  to make sure we don't get unexpected val ? I still see it as some anti-pattern at least for the cases like this.
Moreover I browsed available integrations to understand how widespread construction "if TYPE_CHECKING" is. What I see is only 64 of 1261 existing integrations use it which does not seem so popular. So I prefer not to use it at least while I don't understand the value it provides.
You can check number of usages yourself:
```
cd homeassistant/components
grep "if TYPE_CHECKING" -r | grep -oe "^[^/[:blank:]]*" | uniq -u | wc -l
```

Reply:
The assertion in this case would only really be used for the type checker to pass. I initially also did it in a PR without the if TYPE_CHECKING, but I was told to not use assertions or else use the conditional check for TYPE_CHECKING.

Not sure if there is a better way to make the type checker pass